src/file_loader.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2013-2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib::object::IsA;
+use glib::translate::*;
+use gtk_source_sys;
+use std::boxed::Box;
+use std::ptr;
+use FileLoader;
+
+impl FileLoader {
+    pub fn load_async<
+        P: IsA<gio::Cancellable>,
+        Q: 'static + Fn(i64, i64) + Send,
+        R: 'static + FnOnce(Result<bool, glib::Error>) + Send,
+        S: 'static + FnOnce() + Send,
+    >(
+        &self,
+        io_priority: glib::Priority,
+        cancellable: Option<&P>,
+        progress_callback: Q,
+        ready_callback: R,
+        notify_callback: S,
+    ) {
+        let progress_user_data: Box<(Q, S)> = Box::new((progress_callback, notify_callback));
+        unsafe extern "C" fn load_async_progress_trampoline<
+            Q: 'static + Fn(i64, i64) + Send,
+            S: 'static + FnOnce() + Send,
+        >(
+            current_num_bytes: i64,
+            total_num_bytes: i64,
+            user_data: glib_sys::gpointer,
+        ) {
+            let callbacks: Box<(Q, S)> = Box::from_raw(user_data as *mut _);
```

Comment:
Don't use `from_raw()` here but a simple cast

```suggestion
            let callbacks: &(Q, S) = &*(user_data as *mut (Q, S));
```

Reply:
Same thing in the other places, I won't add it to each

Reply:
OK, will do that. Presumably it only applies to the places where the Box is not to be dropped (one each in `file_loader.rs` and `file_saver.rs`)

Reply:
Yes, this prevents doing the `from_raw()` / `into_raw()` dance and is safer as you can't accidentally forget `into_raw()` in one code branch (which then would cause a double free)